### ISSUE TITLE
Actual alternative to .pipe()?

### ISSUE THREAD

**Issue by JakenVeina** [jquery/jquery#4134]
Obviously, `.then()` is listed as the official replacement for `.pipe()`, however as described in #3325, `.then()` forces execution of the callback onto a new context, even in the case of a promise/deferred that is already resolved, unlike `.pipe()`.

In the following case, `.then()` is not a suitable replacement for `.pipe()`, for the above reason.

```js
var promise = someAsyncOperation()
    .then(function(results)) {
        return processResults(results);
    });

if (promise.state() !== "resolved") {
    $overlay.fadeIn(500);

    promise
        .done(function() {
            $overlay.fadeOut(500);
        });
}

return promise;
```

If `someAsyncOperation()` utilizes `.then()` to filter results anywhere in the call chain, the `$overlay` animations will play unintentionally.

As far as I can tell, there's no alternative in the current library to combine the "immediate callback" behavior of `.done()` with the filtering behavior of `.then()`, except for the deprecated method `.pipe()`. Am I missing something?

**Comment by dmethvin** [comment]
It's been a while, but I think the point of deprecating `.pipe()` was to encourage people to use the Promise/A+ alternatives. At this point it doesn't seem that logical to ever actually *drop* it since we still support the old `Deferred` sync/non-catching methods like `.done()` and `.fail()`. Maybe @gibson042 remembers more about that. Here are some discussions from old tickets:

https://github.com/jquery/jquery/pull/1996

https://github.com/jquery/jquery/pull/1363

**Comment by timmywil** [comment]
@dmethvin is right. `.pipe()` was deprecated before we made `.then` Promises/A+ compliant. However, we left it deprecated because we want to encourage `.then` usage, which is a suitable replacement for most cases. The few cases where `.then` is not an easy replacement for `.pipe` are why we're not removing `.pipe` any time soon.

In other words, the choice here is to un-deprecate `.pipe` or leave deprecated and continue encouraging `.then` usage where we can. We discussed in the meeting and chose the latter. However, I can assure you that deprecation does not always mean removal, and this is one of those cases.
